api/turing/cluster/servicebuilder/router.go

```diff
@@ -199,7 +199,8 @@ func (sb *clusterSvcBuilder) buildRouterEnvs(
 			{Name: envEnricherTimeout, Value: ver.Enricher.Timeout},
 		}...)
 	}
-	if ver.Ensembler != nil && ver.Ensembler.Type == models.EnsemblerDockerType {
+	if ver.Ensembler != nil && ((ver.Ensembler.Type == models.EnsemblerPyFuncType && ver.Ensembler.
+		DockerConfig != nil) || ver.Ensembler.Type == models.EnsemblerDockerType) {
```

Comment:
Actually yes, because should the original deployment process fail due to an image building failure, there wouldn't be an associated Docker config for the pyfunc ensembler specified. Turing API will then attempt to undeploy the current router version, and part of the undeployment process involves calling `buildRouterEnvs`, which is the function which you quoted above. 

If the check for `ver.Ensembler.DockerConfig != nil` didn't happen, we'll end up accessing an invalid memory address when trying to access variables in the `DockerConfig` just shortly afterwards: https://github.com/gojek/turing/blob/deda5df4145c328e235204d8858507379a1fafc3/api/turing/cluster/servicebuilder/router.go#L208

Reply:
Hmm... that is true xD Maybe there was originally a concern that an ensembler config not having the type `docker`, but with a `DockerConfig` getting all the way to that stage? But yeah I don't think that's possible, so I'll just update those checks with that suggestion you have.

Reply:
I'd even consider writing a function to check if it's either a EnsemblerPyFuncType or a EnsemblerDockerType so that it doesn't look that gory.